### PR TITLE
Enable the use of curly brackets in data_filter()

### LINKED ISSUE #216: Programming with `datawizard::data_filter` and character vectors
In the context of my easystats/performance#443 PR, I’ve experienced a difficulty using `datawizard::data_filter`, so I am moving the discussion here.

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(datawizard)
```

With the tidyverse, the recommendation is to use double curly brackets when passing a variable name as argument, e.g.,

``` r
fun.dp <- function(df, var) {
  head(filter(df, {{var}} > 0.5))
}
fun.dp(mtcars, am)
#>                 mpg cyl  disp  hp drat    wt  qsec vs am gear carb
#> Mazda RX4      21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4
#> Mazda RX4 Wag  21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4
#> Datsun 710     22.8   4 108.0  93 3.85 2.320 18.61  1  1    4    1
#> Fiat 128       32.4   4  78.7  66 4.08 2.200 19.47  1  1    4    1
#> Honda Civic    30.4   4  75.7  52 4.93 1.615 18.52  1  1    4    2
#> Toyota Corolla 33.9   4  71.1  65 4.22 1.835 19.90  1  1    4    1
```

In `datawizard`, a naively copy-pasted strategy does not seem to work:

``` r
fun.dw <- function(df, var) {
  head(data_filter(df, filter = {{var}} > 0.5))
}
fun.dw(mtcars, am)
#> Error in {: comparison (6) is possible only for atomic and list types
```

In any case, my needs might be more for passing a character vector. Therefore in `dplyr`, we can use the double square brackets along the `.data` argument:

``` r
fun.dp <- function(df, var) {
  head(filter(df, .data[[var]] > 0.5))
}
fun.dp(mtcars, "am")
#>                 mpg cyl  disp  hp drat    wt  qsec vs am gear carb
#> Mazda RX4      21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4
#> Mazda RX4 Wag  21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4
#> Datsun 710     22.8   4 108.0  93 3.85 2.320 18.61  1  1    4    1
#> Fiat 128       32.4   4  78.7  66 4.08 2.200 19.47  1  1    4    1
#> Honda Civic    30.4   4  75.7  52 4.93 1.615 18.52  1  1    4    2
#> Toyota Corolla 33.9   4  71.1  65 4.22 1.835 19.90  1  1    4    1
```

In `datawizard`, is there any way to achieve the same result?

``` r
fun.dw <- function(df, var) {
  head(data_filter(df, filter = var > 0.5))
}
fun.dw(mtcars, "am")
#> Error in var > 0.5: comparison (6) is possible only for atomic and list types

fun.dw <- function(df, var) {
  head(data_filter(df, filter = deparse(substitute(var)) > 0.5))
}
fun.dw(mtcars, "am")
#>                    mpg cyl disp  hp drat    wt  qsec vs am gear carb
#> Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1    4    4
#> Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1    4    4
#> Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1    4    1
#> Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0    3    1
#> Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0    3    2
#> Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0    3    1
# This doesn't throw an error but doesn't filter correctly

fun.dw <- function(df, var) {
  head(data_filter(df, filter = !!var > 0.5))
}
fun.dw(mtcars, "am")
#> Error in var > 0.5: comparison (6) is possible only for atomic and list types

fun.dw <- function(df, var) {
  head(data_filter(df, filter = .data[[var]] > 0.5))
}
fun.dw(mtcars, "am")
#> Error:
#> ! Can't subset `.data` outside of a data mask context.

fun.dw <- function(df, var) {
  head(data_filter(df, filter = !!!var > 0.5))
}
fun.dw(mtcars, "am")
#> Error in var > 0.5: comparison (6) is possible only for atomic and list types
```

Is there a way to avoid having to do this?

``` r
fun.dw <- function(df, var) {
  df$x <- df[[var]]
  df <- head(data_filter(df, filter = x > 0.5))
  data_remove(df, "x")
}
fun.dw(mtcars, "am")
#>                 mpg cyl  disp  hp drat    wt  qsec vs am gear carb
#> Mazda RX4      21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4
#> Mazda RX4 Wag  21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4
#> Datsun 710     22.8   4 108.0  93 3.85 2.320 18.61  1  1    4    1
#> Fiat 128       32.4   4  78.7  66 4.08 2.200 19.47  1  1    4    1
#> Honda Civic    30.4   4  75.7  52 4.93 1.615 18.52  1  1    4    2
#> Toyota Corolla 33.9   4  71.1  65 4.22 1.835 19.90  1  1    4    1
```

From the documentation I saw that `data_filter` is based on `subset`, which itself has the following warning:

> This is a convenience function intended for use interactively. For programming it is better to use the standard subsetting functions like `[`, and in particular the non-standard evaluation of argument subset can have unanticipated consequences.

Is there a workaround or some way to convert from string to an unevaluated promised expression (or something like that, I don’t really know the language)?

<sup>Created on 2022-08-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### PR DESCRIPTION
**Pull request by strengejacke**
Tries to close #216 

But this is not really working yet... Just different attempts to solve that issue.

### CODE REVIEW

File: R/data_match.R [hunk 1]
```diff
@@ -152,6 +152,31 @@ data_filter <- function(x, filter, ...) {
   if (is.numeric(rows)) {
     out <- x[rows, , drop = FALSE]
   } else {
+
+    # check for curley tags, like "{var} == 1"
+    has_curley <- tryCatch(grepl("{", eval(condition), fixed = TRUE),
+                           error = function(e) FALSE)
```
Comment (strengejacke) [inline R/data_match.R:158]: `eval(condition)` splits the "filter" argument at every `{` and `}`, resulting in a character vector length > 1, however we just want the `filter` argument as on string here.

**Comment by strengejacke** [comment]
seems like `data_wizard(mtcars, filter = {var} != 0)` evaluates the `filter` argument immediately after function call and sets it to `TRUE`, so my whole approach here doesn't seem to work.

**Comment by strengejacke** [comment]
Can we somehow convert this into a draft, so it won't be merged accidentally?

**Comment by IndrajeetPatil** [comment]
Done!

**Comment by etiennebacher** [comment]
[gluestick](https://github.com/coolbutuseless/gluestick/) might be useful here (but I couldn't make it work so far)

**Comment by etiennebacher** [comment]
Much easier to implement thanks to the previous commits this week. All tests pass locally, and here are some examples. Is this what you had in mind?

When vars are in global environment:
``` r
library(datawizard)

var = "mpg"
var2 = "hp"

data_filter(mtcars, "{var} > 30 & {var2} <= 66")
#>                 mpg cyl disp hp drat    wt  qsec vs am gear carb
#> Fiat 128       32.4   4 78.7 66 4.08 2.200 19.47  1  1    4    1
#> Honda Civic    30.4   4 75.7 52 4.93 1.615 18.52  1  1    4    2
#> Toyota Corolla 33.9   4 71.1 65 4.22 1.835 19.90  1  1    4    1
```

<sup>Created on 2022-09-02 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

When vars are passed in a function:
``` r
library(datawizard)

foo <- function(var, var2) {
  data_filter(mtcars, filter = {var} > 30 & {var2} <= 66)
}
foo("mpg", "hp")
#>                 mpg cyl disp hp drat    wt  qsec vs am gear carb
#> Fiat 128       32.4   4 78.7 66 4.08 2.200 19.47  1  1    4    1
#> Honda Civic    30.4   4 75.7 52 4.93 1.615 18.52  1  1    4    2
#> Toyota Corolla 33.9   4 71.1 65 4.22 1.835 19.90  1  1    4    1
```

<sup>Created on 2022-09-02 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by etiennebacher** [comment]
@rempsyc since you opened #216, do you want to check if this PR solves your use cases?

**Comment by rempsyc** [comment]
Wooah, that's amazing! Commit [2c58497](https://github.com/easystats/datawizard/commit/2c584977f9aed4027e6cf08a80c57ffbec25fe43) solved my specific use case for `check_outliers` by allowing complete character string usage (as in your second function example), and this one solves my general use case as described in the PR. This can close #216 now. Thanks!

**Comment by etiennebacher** [comment]
@strengejacke Are you ok with the examples above? Or do you have other use cases to test?

**Comment by strengejacke** [comment]
> @strengejacke Are you ok with the examples above? Or do you have other use cases to test?

I had not many particular use cases in mind, your examples look great and are already an improvement to what we currently have. We should probably - if there's such a case - describe what works and what known issue are (i.e. what should users expect no being working at the moment).

**Pull request merged**